### PR TITLE
Jest integration and initial tests for Header

### PR DESCRIPTION
**Pull request by nam-m**
## Summary of changes
- Installed jest packages and configured project for testing
- Added `href` to Header navigation links and `textDecoration` styles
- Added component and snapshot tests for Header

## Notes
- These tests will serve as templates for developing UI components
- Jest setup and usage doc can be found under [Testing template](https://github.com/OpenBCca/openbc-web/wiki/Testing-template)

### CODE REVIEW

**Comment by SamHuo213** [comment]
Please merge from main.

**Comment by SamHuo213** [comment]
Thanks for your PR,

I have the following comments:

1) jest.config.mjs, are all the comments necessary?
2) Can jest.setup.js be a ts file? What is this file for?
3) What is clsx package?
4) What is the intention of __tests__ folder?
5) is header.snapshot.tsx intended for demo?
6) is index.test.ts intended for demo?
7) I think it should say 'should render header' and describe is the component name and/or function name, on error the print out will be 'header should render header'
8) are the href necessary?

**Comment by nam-m** [comment]
@SamHuo213 thanks for reviewing, here are my answers

>     1. jest.config.mjs, are all the comments necessary?

I think they're helpful to me, maybe i can summarize them better? The comments are from the template in [Nextjs setup guide](https://nextjs.org/docs/pages/building-your-application/optimizing/testing#quickstart-2)

>     2. Can jest.setup.js be a ts file? What is this file for?
I created during a guide for setup using Babel, but we're not using Babel so I will remove this

>     3. What is clsx package?
I was testing sth with it, but didn't need anymore. I will remove then

>     4. What is the intention of **tests** folder?
We can decide on whether to keep the tests here or in each component folder. I guess this is up to discussion

>     5. is header.snapshot.tsx intended for demo?
Yes, this snapshot is for developers to compare their snapshots to the one on `main` repo

>     6. is index.test.ts intended for demo?
Yes, but I will remove this as there are other demo tests

>     7. I think it should say 'should render header' and describe is the component name and/or function name, on error the print out will be 'header should render header'
Good point, will change

>     8. are the href necessary?
`href` is needed to test `toHaveAttribute` in `Header.test.tsx`, I think this will be useful to test the navigation links later

File: src/client-services/github.client-service.ts [hunk 1]
```diff
@@ -1,33 +1,38 @@
...
+  )
+  return response
 }
 
-export async function getRepoInfo(repo_name: string, para: string = ''): Promise<AxiosResponse> {
-  const response = await axiosConfig().get(`/repos/OpenBCca/${repo_name}${para}`);
-  return response;
+export async function getRepoInfo(
+  repo_name: string,
+  para: string
```
Comment (nam-m) [inline src/client-services/github.client-service.ts:32]: @SamHuo213 I got this error
```
Error: Type string trivially inferred from a string literal, remove type annotation.  @typescript-eslint/no-inferrable-types
```
for `para: string = ''`, so I changed it to `para: string` here

**Comment by nam-m** [comment]
@SamHuo213 I have made changes to the code, so it can be reviewed again now

**Comment by SamHuo213** [comment]
> I think they're helpful to me, maybe i can summarize them better? The comments are from the template in [Nextjs setup guide](https://nextjs.org/docs/pages/building-your-application/optimizing/testing#quickstart-2)

I subscribe to self documenting code: https://swimm.io/learn/documentation-tools/tips-for-creating-self-documenting-code

Also these conventions are nice: https://gist.github.com/anichitiandreea/e1d466022d772ea22db56399a7af576b

I ask if these comments are really necessary but I am ok with leaving them :)

> I created during a guide for setup using Babel, but we're not using Babel so I will remove this

Thanks, let's keep things clean.

> I was testing sth with it, but didn't need anymore. I will remove then

Thanks, let's keep things clean.

> We can decide on whether to keep the tests here or in each component folder. I guess this is up to discussion

I am familiar with two conventions:

1) In the same folder. This is typical of JS projects/frameworks
2) Separate Test section with folder structure that mirrors existing structure. This is usually in C# frameworks.

Either works for me, but we should get buy in for our prefer convention.

> Yes, this snapshot is for developers to compare their snapshots to the one on `main` repo

Ok

> Yes, but I will remove this as there are other demo tests

Ok

> Good point, will change

Ok

> `href` is needed to test `toHaveAttribute` in `Header.test.tsx`, I think this will be useful to test the navigation links later

Ok

Usually I prefer not having production code that is only intended for testing but in this case I think it should be ok.

File: src/__tests__/__snapshots__/header.snapshot.tsx.snap [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (SamHuo213) [inline src/__tests__/__snapshots__/header.snapshot.tsx.snap:1]: Is this auto generated code? Can we remove the comment?

File: src/client-services/github.client-service.ts (hunk 1 above)
Comment (SamHuo213) [reply src/client-services/github.client-service.ts:32]: We need to update the lint configurations.

Also, @PooriaT needs to fix the build.

At this time can, you can ignore it I guess.

File: src/client-services/github.client-service.ts [hunk 3]
```diff
@@ -1,33 +1,38 @@
-import axios, { AxiosResponse, AxiosInstance } from 'axios';
+import axios, { AxiosResponse, AxiosInstance } from 'axios'
```
Comment (SamHuo213) [inline src/client-services/github.client-service.ts:1]: Does removing the semi-colon cause lint errors?

After we update the linter, this should cause errors or be auto fixed.

File: src/components/Header/Header.test.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import Header from '@/components/Header/Header'
+import '@testing-library/jest-dom'
+import { render, screen } from '@testing-library/react'
+
+describe('Header', () => {
+  it('renders header', () => {
+    render(<Header />)
+
+    // Query header element
```
Comment (SamHuo213) [inline src/components/Header/Header.test.tsx:9]: Do we need these comments?

File: src/__tests__/__snapshots__/header.snapshot.tsx.snap (hunk 2 above)
Comment (nam-m) [reply src/__tests__/__snapshots__/header.snapshot.tsx.snap:1]: yeah this is auto generated

File: src/client-services/github.client-service.ts (hunk 3 above)
Comment (nam-m) [reply src/client-services/github.client-service.ts:1]: I think eslint auto removes semicolon. I tried putting at the end, but it got removed after saving the file :thinking:

File: src/components/Header/Header.test.tsx (hunk 4 above)
Comment (nam-m) [reply src/components/Header/Header.test.tsx:9]: That comment is for this 
```
// Query header element
const header = screen.getByRole('banner')
```
I can remove it if you think it's unecessary

File: src/client-services/github.client-service.ts (hunk 3 above)
Comment (SamHuo213) [reply src/client-services/github.client-service.ts:1]: Interesting, after we update our Eslint rules, it should auto add semi-colons.

This is fine for now.

File: src/components/Header/Header.test.tsx (hunk 4 above)
Comment (SamHuo213) [reply src/components/Header/Header.test.tsx:9]: I think this is a good example of comments being noise.

To be it's obvious you are querying for the header by the code.

Comment (nam-m) [reply src/components/Header/Header.test.tsx:9]: alright, will remove this and other comments below :+1:

File: .eslintrc.cjs [hunk 5]
```diff
@@ -34,13 +37,14 @@ module.exports = {
...
-    '@typescript-eslint/no-unsafe-assignment': 'warn',
+    '@typescript-eslint/no-unsafe-assignment': 'off',
     '@typescript-eslint/no-unsafe-call': 'warn',
     '@typescript-eslint/no-unsafe-member-access': 'warn',
     '@typescript-eslint/no-non-null-assertion': 'warn',
     '@typescript-eslint/no-unnecessary-condition': 'warn',
     '@typescript-eslint/ban-types': 'warn',
     '@typescript-eslint/no-empty-interface': 'warn',
     '@typescript-eslint/no-unused-vars': 'warn',
+    '@typescript-eslint/no-empty-function': 'warn',
```
Comment (umsu2) [inline .eslintrc.cjs:48]: we shouldn't be changing linting rules unless there is a good reason to. once linting is set, it's like the standard the team needs to obey by.

Comment (SamHuo213) [reply .eslintrc.cjs:48]: I agree,

To give you context, the previous rules had issues and that is why @nam-m , needed to temporary fix it here.

@nam-m , make sure to pull the latest into your branch and also ensure you remove any linting changes.

Comment (nam-m) [reply .eslintrc.cjs:48]: Gotcha, will update with latest lint rules

**Comment by nam-m** [comment]
@umsu2 @SamHuo213 I have done the following:
- Merged latest main
- Added semicolon to test files
- Installed testing packages in package.json and combined package-lock.json
- Updated header & header snapshot tests

It took me some time going through all the changes in package-lock.json to include/exclude packages in vscode, wondering if there's a quick way to do this?

**Comment by SamHuo213** [comment]
> @umsu2 @SamHuo213 I have done the following:
> 
> * Merged latest main
> * Added semicolon to test files
> * Installed testing packages in package.json and combined package-lock.json
> * Updated header & header snapshot tests
> 
> It took me some time going through all the changes in package-lock.json to include/exclude packages in vscode, wondering if there's a quick way to do this?

package-lock.json is a autogenerated file. The best way was to delete it and do an npm install.

File: .eslintrc.cjs [hunk 6]
```diff
@@ -1,4 +1,7 @@
 module.exports = {
+  env: {
```
Comment (SamHuo213) [inline .eslintrc.cjs:2]: Do you need this?

**Comment by SamHuo213** [comment]
> > @umsu2 @SamHuo213 I have done the following:
> > 
> > * Merged latest main
> > * Added semicolon to test files
> > * Installed testing packages in package.json and combined package-lock.json
> > * Updated header & header snapshot tests
> > 
> > It took me some time going through all the changes in package-lock.json to include/exclude packages in vscode, wondering if there's a quick way to do this?
> 
> package-lock.json is a autogenerated file. The best way was to delete it and do an npm install.

@nam-m , can you please make sure to do this please :).

File: .eslintrc.cjs (hunk 6 above)
Comment (nam-m) [reply .eslintrc.cjs]: I had it for eslint-plugin-jest, not currently needed so I removed it :+1:

**Comment by nam-m** [comment]
> > > @umsu2 @SamHuo213 I have done the following:
> > > 
> > > * Merged latest main
> > > * Added semicolon to test files
> > > * Installed testing packages in package.json and combined package-lock.json
> > > * Updated header & header snapshot tests
> > > 
> > > It took me some time going through all the changes in package-lock.json to include/exclude packages in vscode, wondering if there's a quick way to do this?
> > 
> > 
> > package-lock.json is a autogenerated file. The best way was to delete it and do an npm install.
> 
> @nam-m , can you please make sure to do this please :).

done

**Review by SamHuo213 — APPROVED** [review]

**Review by umsu2 — APPROVED** [review]

**Pull request merged**